### PR TITLE
buildkit: add image

### PR DESCRIPTION
**Pull request by wlynch**
### Image Size

- [x] The Image is smaller in size than its common public counterpart.
- [ ] The Image is larger in size than its common public counterpart (please explain in the notes).

Notes:

Image | Size
-- | --
moby/buildkit:latest | 168MB
cgr.dev/chainguard/buildkit:latest | 69.2MB

(though slightly misleading since our image is not installing everything 1:1)

### Image Vulnerabilities

- [ ] The Grype vulnerability scan returned 0 CVE(s).
- [x] The Grype vulnerability scan returned > 0 CVE(s) (please explain in the notes).

Notes:

```
NAME                                   INSTALLED  FIXED-IN  TYPE       VULNERABILITY        SEVERITY 
github.com/cyphar/filepath-securejoin  v0.2.3     0.2.4     go-module  GHSA-6xv5-86q9-7xr8  Medium
```

Needs to be fixed in the Wolfi package (will bump in another PR): https://github.com/wolfi-dev/os/pull/6618

### Basic Testing - K8s cluster

- [ ] The container image was successfully loaded into a kind cluster.
- [x] The container image could not be loaded into a kind cluster (please explain in the notes).

Notes:

### Basic Testing - Package/Application

- [x] The application is accessible to the user/cluster/etc. after start-up.
- [ ] The application is not accessible to the user/cluster/etc. after start-up. (please explain in the notes).

Notes:

### Helm

- [ ] A Helm chart has been provided and the container image can be used with the chart.  If needed, please add a -compat package to close any gaps with the public helm chart.
- [ ] A Helm chart has been provided and the container image is not working with the chart (please explain in the notes).
- [x] A Helm chart was not provided.

Notes:

Script based test

### Processor Architectures

- [ ] The image was built and tested for x86_64.
- [ ] The image could not be built for x86_64 (please explain in the notes).
- [ ] The image was built and tested for aarch64.
- [ ] The image could not be built for aarch64. (please explain in the notes).

Notes:

### Functional Testing + Documentation

- [x] Functional tests have been included and the tests are passing.  All tests have been documnted in the notes section.

Notes:

### Environment Testing + Documentation

- [x] There has not been a request and/or there is no indication that this image needs tested on a public cloud provider.
- [ ] The container image has been tested successfully on a public cloud provider (AWS, GCP, Azure).
- [ ] The container image has not been tested successfully on a public cloud provider (AWS, GCP, Azure) (please explain in the notes).

Notes:

### Version

- [x] The package version is the latest version of the package.  The latest tag points to this version.
- [ ] The package version is the not the latest version of the package (please explain in the notes).

Notes:

### Dev Tag Availability

- [x] There is a dev tag available that includes a shell and apk tools (by depending on 'wolfi-base')
- [ ] There is not a dev tag available that includes a shell and apk tools (by depending on 'wolfi-base') (please explain in the notes).

Notes:

### Access Control + Authentication

- [ ] The image runs as `nonroot` and GID/UID are set to 65532 or upstream default
- [x] Alternatively the username and GID/UID may be a commonly used one from the ecosystem e.g: postgres
- [ ] The image requires a non-standard username or non-standard GID/UID (please explain in the notes).

Notes:

Default image runs as root. (WIP rootless will run as non-root)

### ENTRYPOINT

- [x] applications/servers/utilities set to call main program with no arguments e.g. [redis-server]
- [ ] applications/servers/utilities not set to call main program with no arguments e.g. [redis-server] (please explain in the notes)
- [ ] base images leave empty.
- [ ] base image and not empty (please explain in the notes).
- [ ] dev variants is set to entrypoint script that falls back to system.
- [ ] dev variants is not set to entrypoint script that falls back to system (please explain in the notes).

### CMD

- [ ] For server applications give arguments to start in daemon mode (may be empty)
- [x] For utilities/tooling bring up help e.g. `–help`
- [ ] For base images with a shell, call it e.g. [/bin/sh]

Notes:

Matches upstream entrypoint + args

### Environment Variables

- [x] Environment variables added.
- [ ] Environment variables not added and not required.

### SIGTERM

- [x] The image responds to SIGTERM (e.g., `docker kill $(docker run -d --rm cgr.dev/chainguard/nginx)`)

### Logs

- [x] Error logs write to stderr and normal logs to stdout. Logs DO NOT write to file.

### Documentation - README

- [x] A README file has been provided and it follows the README template.

### CODE REVIEW

File: images/buildkit/config/main.tf
Comment (mattmoor) [inline images/buildkit/config/main.tf:17]: I think this works:
```suggestion
```

File: images/buildkit/config/main.tf
Comment (mattmoor) [inline images/buildkit/config/main.tf:18]: perhaps instead of instantiating accts twice and optionally using one we can have the ternary on `run-as = rootless ? 65532 : 0`?

File: tflib/publisher/main.tf
Comment (mattmoor) [inline tflib/publisher/main.tf:170]: Hmm this seems out of place, why do you need this?

File: images/buildkit/main.tf
Comment (mattmoor) [inline images/buildkit/main.tf:71]: should we aim for `rootless` by default and ship `:root` tags, like we do for `git`?

File: images/buildkit/main.tf
Comment (mattmoor) [inline images/buildkit/main.tf:53]: I'd directly `oci_tag` these.  We can add a downstream rebuild if we want version tags.

File: images/buildkit/config/main.tf
Comment (wlynch) [reply images/buildkit/config/main.tf:18]: This is actually used (left over from something I was trying before). Removed!

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf:71]: I was trying to stay consistent with upstream tags: https://hub.docker.com/r/moby/buildkit.

I'll throw this back to you: how much are we okay with deviating from upstream tags?

File: tflib/publisher/main.tf
Comment (wlynch) [reply tflib/publisher/main.tf:170]: This is skew catch up from images-internal. 🤷

I think I can drop this though with the oci_tag change.

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: Done

File: images/buildkit/config/main.tf
Comment (wlynch) [reply images/buildkit/config/main.tf]: Removed

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: more context: Buildkit might be somewhat different than other images since root/privileged is it's "default" mode. 
To invoke rootless you need a [whole other set of arguments](https://github.com/moby/buildkit/blob/master/docs/rootless.md#containerized-deployment) which means that it's not a drop in replacement for docs that are using `:latest` assuming a rootfull setup.

Comment (mattmoor) [reply images/buildkit/main.tf]: I think it's fine to deviate as long as we have a good reason.

**Comment by mattmoor** [comment]
LGTM, lmk when tests are passing.

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: Renamed to `latest-root`!

File: images/buildkit/main.tf
Comment (mattmoor) [inline images/buildkit/main.tf:35]:
```suggestion
  depends_on = [module.test-latest-root]
```

File: images/buildkit/main.tf
Comment (mattmoor) [inline images/buildkit/main.tf:41]:
```suggestion
  depends_on = [module.test-latest-root]
```

File: images/buildkit/tests/02-build.sh
Comment (mattmoor) [inline images/buildkit/tests/02-build.sh:12]: per your TODOs about `awk` etc: Would it make more sense for us to make this the entrypoint and include `busybox` (which provides most posix utils)?

File: images/buildkit/tests/main.tf
Comment (mattmoor) [inline images/buildkit/tests/main.tf:14]: We've been starting to forgo these superficial checks when we have more substantial testing, and I think the below clearly suffices!

File: images/buildkit/tests/main.tf
Comment (mattmoor) [inline images/buildkit/tests/main.tf:18]: The `02-` prefix is a vestige of the bygone era of `monopod` crafting shell invocations where this affected ordering.  I've been encouraging folks to just drop the prefix these days.

**Comment by mattmoor** [comment]
Very cool to see this all working @wlynch 🤩

File: images/buildkit/tests/02-build.sh
Comment (wlynch) [reply images/buildkit/tests/02-build.sh:12]: We could include busybox / bash - I was holding off on it because I wasn't sure how minimal we wanted to make the non-dev variant. These are only needed because of how buildkit recommends launching daemonless via the script. AFAIK these aren't needed when using this as a daemon, and it's probably okay to say "if you want to use daemonless use the dev variant"? We can also wait for any feedback on this and add it later.

daemonless is not the default mode for buildkit, so I'd say it doesn't make sense to make this the entrypoint. We're using it in tests because I didn't know how we can install script dependencies in tests (since we need the buildctl CLI). Running it daemonless allows us to encapsulate everything in a single docker command.

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: Done.

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: Done.

**Comment by wlynch** [comment]
CI will fail for the latest iteration - waiting on https://github.com/wolfi-dev/os/pull/7047

File: images/buildkit/main.tf
Comment (mattmoor) [inline images/buildkit/main.tf:65]: I think we probably want this to use `${path.module}/tests/build-rootless.sh`

File: images/buildkit/main.tf
Comment (mattmoor) [inline images/buildkit/main.tf:32]: same as below

File: images/buildkit/tests/main.tf
Comment (mattmoor) [inline images/buildkit/tests/main.tf:17]: per comments above, I'd drop `${path.module}/` here

File: images/buildkit/config/main.tf
Comment (mattmoor) [inline images/buildkit/config/main.tf:31]: Can you add a comment explaining this?  Seems odd and it'd be good to record the context you have for "tomorrow Billy (or Matt or Jason)" :)

**Comment by mattmoor** [comment]
I reran the leg now that the Wolfi PR landed, and it still failed.  Haven't cracked logs.

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: Done.

File: images/buildkit/main.tf
Comment (wlynch) [reply images/buildkit/main.tf]: Done.

File: images/buildkit/tests/main.tf
Comment (wlynch) [reply images/buildkit/tests/main.tf]: Done.

File: images/buildkit/config/main.tf
Comment (wlynch) [reply images/buildkit/config/main.tf]: Done.

**Comment by wlynch** [comment]
> I reran the leg now that the Wolfi PR landed, and it still failed. Haven't cracked logs.

Waiting on https://github.com/wolfi-dev/os/pull/7106 to get merged

File: images/buildkit/tests/main.tf
Comment (mattmoor) [inline images/buildkit/tests/main.tf:18]: NB: at this point it seems somewhat silly to even have a module for this 🤣

**Review by mattmoor — APPROVED** [review]

**Comment by mattmoor** [comment]
Blah, I reran, but I don't think I waited for the post-submit build 🤦

**Comment by wlynch** [comment]
The rabbithole continues...

```
│ [rootlesskit:parent] error: failed to setup UID/GID map: newuidmap 21 [0 65532 1 1 100000 65536] failed: : exec: "newuidmap": executable file not found in $PATH
```

I'm going to try and get a local repo set up for testing so I don't need to keep RTing on package update PRs. 😢

File: images/buildkit/tests/main.tf
Comment (wlynch) [reply images/buildkit/tests/main.tf]: Removed.

**Review by mattmoor — APPROVED** [review]

**Pull request merged**
